services/templates/service.go

```diff
@@ -62,23 +78,110 @@ func (s service) AddTemplate(c context.Context, template *entities.Template) err
 	return nil
 }
 
-func New() Service {
-	return &service{}
+func (s service) UpdateTemplate(c context.Context, template *entities.Template) error {
+	// parse string to validate template params
+	_, err := mustache.ParseString(template.HTMLPart)
+	if err != nil {
+		return ErrParseHTMLPart
+	}
+	// parse string to validate template params
+	_, err = mustache.ParseString(template.TextPart)
+	if err != nil {
+		return ErrParseTextPart
+	}
+	// parse string to validate template params
+	_, err = mustache.ParseString(template.SubjectPart)
+	if err != nil {
+		return ErrParseSubjectPart
+	}
+
+	s3Input := &s3.PutObjectInput{
+		Bucket: aws.String(templatesBucket),
+		Key:    aws.String(fmt.Sprintf("%d/%d", template.UserID, template.ID)),
+		Body:   bytes.NewReader([]byte(template.HTMLPart)),
+	}
+
+	_, err = s.s3.PutObject(s3Input)
+	if err != nil {
+		return fmt.Errorf("upload template: put s3 object: %w", err)
+	}
+
+	err = s.db.UpdateTemplate(template)
+	if err != nil {
+		return fmt.Errorf("update template: %w", err)
+	}
+
+	return nil
 }
 
 // GetTemplate returns the template with given template id and user id
 func (s service) GetTemplate(c context.Context, templateID int64, userID int64) (*entities.Template, error) {
-	template, err := storage.GetTemplate(c, templateID, userID)
+	template, err := s.db.GetTemplate(templateID, userID)
 	if err != nil {
 		return nil, fmt.Errorf("get template: %w", err)
 	}
 
-	html, err := s3.GetHTMLTemplate(c, userID, template.Name)
+	obj, err := s.s3.GetObject(&s3.GetObjectInput{
```

Comment:
Check out the example for the method here: https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.GetObject

```golang
svc := s3.New(session.New())
input := &s3.GetObjectInput{
    Bucket: aws.String("examplebucket"),
    Key:    aws.String("HappyFace.jpg"),
}

result, err := svc.GetObject(input)
if err != nil {
    if aerr, ok := err.(awserr.Error); ok {
        switch aerr.Code() {
        case s3.ErrCodeNoSuchKey:
            fmt.Println(s3.ErrCodeNoSuchKey, aerr.Error())
        case s3.ErrCodeInvalidObjectState:
            fmt.Println(s3.ErrCodeInvalidObjectState, aerr.Error())
        default:
            fmt.Println(aerr.Error())
        }
    } else {
        // Print the error, cast err to awserr.Error to get the Code and
        // Message from an error.
        fmt.Println(err.Error())
    }
    return
}

fmt.Println(result)
```

We can have a finer-grained error messages to properly log them maybe. Also, do we need to close the body here?

Reply:
@FilipNikolovski I made some changes you can review them